### PR TITLE
Adds filter for duplicate transactions before mining

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

Adds filter for duplicate transactions before mining

### CODE REVIEW

File: ledger/src/state/ledger.rs [hunk 1]
```diff
@@ -558,8 +558,31 @@ impl<N: Network> LedgerState<N> {
         let amount = Block::<N>::block_reward(block_height);
         let coinbase_transaction = Transaction::<N>::new_coinbase(recipient, amount, rng)?;
 
+        // Filter the transactions to ensure they are new, and append the coinbase transaction.
+        // TODO (howardwu): Improve the performance and design of this.
+        let mut transactions: Vec<Transaction<N>> = transactions
+            .iter()
+            .filter(|transaction| {
```
Comment (raychu86) [inline ledger/src/state/ledger.rs:565]: Would a `retain` call be more efficient than a `filter`?

File: ledger/src/state/ledger.rs [hunk 2]
```diff
@@ -558,8 +558,31 @@ impl<N: Network> LedgerState<N> {
...
+                for commitment in transaction.commitments() {
+                    if let Ok(true) = self.contains_commitment(commitment) {
+                        return false;
+                    }
+                }
+
+                true
+            })
+            .cloned()
+            .collect();
```
Comment (raychu86) [inline ledger/src/state/ledger.rs:581]: We should craft the coinbase transaction here and include the transaction fees in the reward.

File: ledger/src/state/ledger.rs (hunk 1 above)
Comment (ljedrz) [reply ledger/src/state/ledger.rs:565]: `transactions` is just a slice, though, so it can't be `retain`ed

**Comment by raychu86** [comment]
Because we just filter out transactions with existing commitments and serial numbers, a transaction could live in the memory pool forever, which seems to open us up to spam attacks. We may want to periodically clean the mempool or do it during block mining.

**Review by raychu86 — APPROVED** [review]

File: ledger/src/state/ledger.rs (hunk 2 above)
Comment (raychu86) [reply ledger/src/state/ledger.rs:581]: Will add in a followup PR.

**Pull request merged**